File: adt2unit/_4op_vol_valid_chan.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// chan: 1..20
/*static*/ bool _4op_vol_valid_chan (uint8_t chan)
{
  chan--;

  uint32_t _4op_flag = _4op_data_flag (chan + 1);

  return    (_4op_flag & 1)              // 1 bit:  A0
         &&  vol4op_lock[chan]
         && (_4op_flag & (0xFFL << 11))  // 8 bits: E7-E0
         && (_4op_flag & (0xFFL << 19)); // 8 bits: F7-F0
}

File: adt2unit/pas/_4op_vol_valid_chan.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function _4op_vol_valid_chan(chan: Byte): Boolean;

var
  _4op_flag: Dword;

begin
  _4op_flag := _4op_data_flag(chan);
  _4op_vol_valid_chan := BOOLEAN(_4op_flag AND 1) and vol4op_lock[chan] and
                     (BYTE(_4op_flag SHR 11) <> 0) and
                     (BYTE(_4op_flag SHR 19) <> 0);
end;
